### PR TITLE
NEUBURG-2: Attach MongoDB

### PR DESCRIPTION
**Pull request by NKlug**
Somethings still need to be changed, like the explicit form(s) for the cities. I also would appreciate that we add some tests eventually.
If you want to test the DB, you should read the ReadMe - Setup :)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,4 +3,6 @@
 /npm-debug.log
```
Comment (maxammann) [inline .gitignore:3]: There is a super cool intellij plugin called .gitignore. It generates the a complete git ignore which adds the ignores for:
* Node
* Intellij
etc.

Maybe generate one with this.

File: package.json [hunk 2]
```diff
@@ -16,24 +17,26 @@
...
-    "nodemailer": "^4.6.0",
     "express-validator": "^5.0.2",
+    "http-status-codes": "1.3.0",
+    "mongodb": "^3.0.4",
+    "mongoose": "^5.0.11",
     "morgan": "^1.9.0",
-    "pug": "^2.0.0-rc.4",
-    "http-status-codes": "1.3.0"
+    "nodemailer": "^4.6.0",
+    "pug": "^2.0.0-rc.4"
```
Comment (maxammann) [inline package.json:26]: Any reason for choosing the rc over the stable?

File: package.json [hunk 3]
```diff
@@ -16,24 +17,26 @@
...
   },
   "devDependencies": {
     "babel-cli": "^6.26.0",
     "babel-core": "^6.26.0",
     "babel-eslint": "^8.2.2",
     "babel-preset-env": "^1.6.1",
+    "es6-template-string-loader": "^2.0.0",
```
Comment (maxammann) [inline package.json:33]: This is a webpack loader isn't it?

File: src/api/offers.js [hunk 4]
```diff
@@ -16,23 +17,31 @@ const validateMiddleware = (req, res, next) => {
 export default ({offerService}) => {
   const router = new Router()
 
-  router.get('/getAll', (req, res) => {
-    res.json(offerService.getAllOffers())
+  router.get('/getAll', async (req, res) => {
```
Comment (maxammann) [inline src/api/offers.js]: I'd like a route like `/offers` a lot more

File: src/api/offers.js [hunk 5]
```diff
@@ -16,23 +17,31 @@ const validateMiddleware = (req, res, next) => {
...
+      return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json(e)
+    }
   })
 
   router.put('/', [
     body('email').isEmail().trim().normalizeEmail(),
     body('duration').isInt().toInt().custom(value => [3, 7, 14, 30].includes(value)),
+    body('formData').exists(),
     validateMiddleware,
     async (req, res) => {
```
Comment (maxammann) [inline src/api/offers.js]: I'd propose to write `request` and `result` always.

File: src/api/offers.js [hunk 6]
```diff
@@ -47,15 +56,16 @@ export default ({offerService}) => {
...
-      const offer = offerService.findOfferByToken(token)
-      if (!offer) {
-        return res.status(HttpStatus.NOT_FOUND).json('No such offer')
-      } else if (offer.isExpired() || offer.deleted) {
-        return res.status(HttpStatus.GONE).json('Offer not available')
-      }
-
       try {
+        const {token} = matchedData(req)
+        const offer = new Offer(await offerService.getOfferByToken(token))
```
Comment (maxammann) [inline src/api/offers.js]: `getOfferByToken` should already return an `Offer`, right?

File: src/api/offers.js [hunk 7]
```diff
@@ -47,15 +56,16 @@ export default ({offerService}) => {
...
+
+        if (!offer) {
+          return res.status(HttpStatus.NOT_FOUND).json('No such offer')
```
Comment (maxammann) [inline src/api/offers.js]: Does return a JSON object? If yes do we have a defined format for error codes? Like:
```
{
    "message": "Error message",
    "code": 55
    "additional_data": {}
}
```

File: src/db.js [hunk 8]
```diff
@@ -1,4 +1,7 @@
+import mongoose from 'mongoose'
+
 export default callback => {
-  // connect to a database if needed, then pass it to `callback`:
-  callback()
+  mongoose.connect('mongodb://localhost/livingDB')
```
Comment (maxammann) [inline src/db.js:4]: Make this url configurable. Maybe also add come commandline parsing to set the config.

https://www.npmjs.com/package/commander

File: src/services/MailService.js [hunk 9]
```diff
@@ -1,7 +1,9 @@
 import {createTransport} from 'nodemailer'
-import smptConfig from '../smptConfig'
+import smtpConfig from '../smtpConfig'
 import {compileFile} from 'pug'
 
+const develop = true
```
Comment (maxammann) [inline src/services/MailService.js:5]: This should be a constant which can be used in the whole app and also settable by command line. Usually node probgramms use the environment variable `NODE_ENV=dev` for this.

File: src/services/OfferService.js [hunk 10]
```diff
@@ -1,58 +1,55 @@
...
+
+    const form = new forms[city](formData)
     const offer = new Offer({
-      id,
-      email,
-      city,
-      formData,
+      email: email,
+      city: city,
       expirationDate: Date.now() + duration * MILLISECONDS_IN_A_DAY,
```
Comment (maxammann) [inline src/services/OfferService.js:17]: Does it make sense to use moment in the backend? If yes create a task for this.

File: src/api/offers.js [hunk 11]
```diff
@@ -16,23 +17,31 @@ const validateMiddleware = (req, res, next) => {
...
-  router.get('/', (req, res) => {
-    res.json(offerService
-      .getActiveOffers(req.city)
-      .map(offer => offer.formData))
+  router.get('/', async (req, res) => {
+    try {
+      const queryResult = await offerService.getActiveOffers(req.city)
+      return res.json(queryResult)
+    } catch (e) {
+      return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json(e)
```
Comment (maxammann) [inline src/api/offers.js]: :warning: This does leak internal errors and stack traces afaik. This should not happend. This error can be logged to the stdout but shouldn't be exposed to the clients.

File: package.json [hunk 12]
```diff
@@ -16,24 +17,26 @@
...
     "eslint": "^4.18.1",
     "eslint-config-standard": "^11.0.0",
     "eslint-plugin-import": "^2.9.0",
     "eslint-plugin-node": "^6.0.1",
     "eslint-plugin-promise": "^3.6.0",
     "eslint-plugin-standard": "^3.0.1",
-    "es6-template-string-loader": "^2.0.0",
     "nodemon": "^1.15.1"
```
Comment (maxammann) [inline package.json:40]: Is there something like hot-code reloading? Would be very cool imo.

Comment (maxammann) [reply package.json:40]: Does not belong to this PR thoug.

File: package.json [hunk 13]
```diff
@@ -16,24 +17,26 @@
...
+    "pug": "^2.0.0-rc.4"
   },
   "devDependencies": {
     "babel-cli": "^6.26.0",
     "babel-core": "^6.26.0",
     "babel-eslint": "^8.2.2",
     "babel-preset-env": "^1.6.1",
+    "es6-template-string-loader": "^2.0.0",
     "eslint": "^4.18.1",
     "eslint-config-standard": "^11.0.0",
```
Comment (maxammann) [inline package.json:35]: Can we import our eslint config? Or is it already? Does not belong to this PR thoug.

File: package.json [hunk 14]
```diff
@@ -3,6 +3,7 @@
   "version": "0.0.1",
```
Comment (maxammann) [inline package.json:3]: We should add a minimum node version here. Not all the node versions support the features like string templates or es6 exports.

**Review by maxammann — CHANGES_REQUESTED** [review]

File: src/api/offers.js (hunk 4 above)
Comment (michael-markl) [reply src/api/offers.js]: this is a route for development only, will be removed when we're approaching production builds

File: src/api/offers.js (hunk 11 above)
Comment (michael-markl) [reply src/api/offers.js]: Yup that's also Dev-only, I'll create a task to remove these

File: package.json (hunk 13 above)
Comment (michael-markl) [reply package.json:35]: It's already imported, but also already outdated again :D
I'll create a task with low prio ^^

File: src/api/offers.js (hunk 7 above)
Comment (michael-markl) [reply src/api/offers.js]: Not yet, should we establish one?

File: src/services/OfferService.js (hunk 10 above)
Comment (michael-markl) [reply src/services/OfferService.js:17]: Rather not, we probably would have to care more about converting from and to the moment objects 🤔

File: src/api/offers.js (hunk 5 above)
Comment (michael-markl) [reply src/api/offers.js]: Yes, that's probably better. Can you do it in this task @NKlug ? :D

File: src/models/Offer.js [hunk 15]
```diff
@@ -1,21 +1,50 @@
...
+  },
+  hashedToken: {
+    type: String,
+    required: [true, 'Missing hashed token'],
+    unique: true // Create a mongodb unique index
+  },
+  formData: {
+    type: mongoose.Schema.Types.ObjectId,
+    ref: Array.from(Object.keys(forms)), // Must be one of the forms defined in forms.js
+    required: [true, 'Missing form data']
```
Comment (michael-markl) [inline src/models/Offer.js:42]: Do we have to write these error messages (for required fields) ourselves? Is mongoose able to create these generic messages itself? 🤔

File: src/models/Offer.js [hunk 16]
```diff
@@ -1,21 +1,50 @@
...
+    default: Date.now
+  },
+  hashedToken: {
+    type: String,
+    required: [true, 'Missing hashed token'],
+    unique: true // Create a mongodb unique index
+  },
+  formData: {
+    type: mongoose.Schema.Types.ObjectId,
+    ref: Array.from(Object.keys(forms)), // Must be one of the forms defined in forms.js
```
Comment (michael-markl) [inline src/models/Offer.js:41]: Can we link the formRef to the city?
If I understand it correctly, currently the city could be 'augsburg' and the formData-ref could be 'neuburg' right?

**Review by michael-markl — CHANGES_REQUESTED** [review]

File: src/api/offers.js (hunk 5 above)
Comment (NKlug) [reply src/api/offers.js]: I'll do it.

File: src/api/offers.js (hunk 11 above)
Comment (NKlug) [reply src/api/offers.js]: With mongoose comes the cool option to set custom error messages for every item in a form. I already partially implemented this, but I think they are not returned atm. We could also return these messages here.

File: src/models/Offer.js (hunk 15 above)
Comment (NKlug) [reply src/models/Offer.js:42]: As far as I know mongoose is able to tell at which item something is wrong, so i.e. that something is wrong with the email. Writing custom error messages is optional .

File: src/models/Offer.js (hunk 16 above)
Comment (NKlug) [reply src/models/Offer.js:41]: This is just the static schema definition; a list of all possible entries that could ever be in this field. I took care of the situation you described in createOffer(). Have a look at [this](https://github.com/Integreat/neuburg-backend/blob/a6b38f1a2e812c3e1933959bf1121e3a6ffe2df6/src/services/OfferService.js#L13 ) :)

File: src/api/offers.js (hunk 6 above)
Comment (NKlug) [reply src/api/offers.js]: Makes sense, I'll have a look at it.

File: src/services/MailService.js (hunk 9 above)
Comment (NKlug) [reply src/services/MailService.js:5]: You are right, I was just looking for a quick way to prevent emails from being sent :)

Comment (NKlug) [reply src/services/MailService.js:5]: Maybe that's a thing that should be done in another task, I'll create one.

File: .gitignore (hunk 1 above)
Comment (NKlug) [reply .gitignore:3]: I created a task for this.

File: src/api/offers.js (hunk 11 above)
Comment (NKlug) [reply src/api/offers.js]: I added this to a task.

File: package.json (hunk 14 above)
Comment (NKlug) [reply package.json:3]: Added this to a task.

File: src/db.js (hunk 8 above)
Comment (NKlug) [reply src/db.js:4]: Created a task for commandline parsing.

**Review by michael-markl — APPROVED** [review]
